applications/OptimizationApplication/python_scripts/responses/response_function.py

```diff
@@ -1,27 +1,115 @@
 from abc import ABC, abstractmethod
+from typing import Union
 import KratosMultiphysics as Kratos
+import KratosMultiphysics.OptimizationApplication as KratosOA
 from KratosMultiphysics.OptimizationApplication.utilities.union_utilities import SupportedSensitivityFieldVariableTypes
 
 class ResponseFunction(ABC):
-    @classmethod
+    """Base response function.
+
+    This reponse function is the base response function. This is assumed to have following responsibilities.
+        1. CalculateValue for a fresh design. (@see CalculateValue)
+        2. CalculateSensitivity for a fresh design (@see CalculateSensitivity)
+
+    This response should only work on one model part. Hence, if multiple model parts required then,
+    a single model part should be created using Kratos.ModelPartOperationUtilities.
+    """
+    def __init__(self, response_name: str) -> None:
+        self.__name = response_name
+
+    def GetName(self) -> str:
+        return self.__name
+
     @abstractmethod
-    def GetSensitivityFieldVariables(cls) -> 'list[SupportedSensitivityFieldVariableTypes]':
+    def Initialize(self) -> None:
+        """Initializes the response.
+
+        This method initializes the response. This is only called once in the whole optimization process.
+
+        """
         pass
 
-    def Initialize(self) -> None:
+    @abstractmethod
+    def Check(self) -> None:
+        """Checks the response.
+
+        This method checks the response. This is only called once in the whole optimization process.
+
+        """
         pass
 
+    @abstractmethod
     def Finalize(self) -> None:
+        """Finalizes the response.
+
+        This method finalizes the response. This is only called once in the whole optimization process.
+
+        """
         pass
 
     @abstractmethod
-    def Check(self) -> None:
+    def GetDependentPhysicalKratosVariables(self) -> 'list[SupportedSensitivityFieldVariableTypes]':
+        """Returns all the dependent physical kratos variables of the response.
+
+        This method should return all the dependent physical variables of the response to make sure
+        that sensitivities w.r.t. those variables are requested from the response. If sensitivities
+        w.r.t. other variables are required from this response, they are ASSUMED TO BE ZERO.
+
+        Please return all the dependent physical variables, eventhough the gradient computation is not yet implemented
+        for some to avoid future bugs.
+
+        Returns:
+            list[SupportedSensitivityFieldVariableTypes]: All dependent physical variables of the response.
+        """
```

Comment:
@RezaNajian Please think about whether we assume zero sensitivities for physical variables which are not implemented where the control have partial overlap, or should the developer give all the influencial physical variables of the response, so then requested physical gradients and allowed gradients can be cross checked return proper zero senstivities for non-influential physical variables.

```suggestion
        """Returns the implemented dependent physical kratos variables of the response.

        This method should return the implemented dependent physical variables of the response to make sure
        that sensitivities w.r.t. those variables are requested from the response. If sensitivities
        w.r.t. other variables are required from this response, they are ASSUMED TO BE ZERO.

        Please return all the dependent physical variables, eventhough the gradient computation is not yet implemented
        for some to avoid future bugs.

        Returns:
            list[SupportedSensitivityFieldVariableTypes]: All dependent physical variables of the response.
        """
```